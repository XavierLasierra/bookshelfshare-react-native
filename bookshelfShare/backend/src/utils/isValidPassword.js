const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
