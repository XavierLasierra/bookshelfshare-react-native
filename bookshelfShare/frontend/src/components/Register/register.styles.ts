import {StyleSheet} from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  topContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    height: 100,
    paddingHorizontal: 30,
  },
  registerContainer: {
    backgroundColor: stylesConstants.colors.light,
    borderTopRightRadius: 40,
    padding: stylesConstants.padding.normal,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'space-between',
    height: 600,
  },
  registerButton: {
    marginTop: 20,
  },
  invalidPasswordText: {
    paddingHorizontal: 5,
  },
});
