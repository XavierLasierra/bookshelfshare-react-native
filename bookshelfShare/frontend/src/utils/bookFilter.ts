export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
