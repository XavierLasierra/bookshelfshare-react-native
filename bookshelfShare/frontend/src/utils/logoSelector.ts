import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
