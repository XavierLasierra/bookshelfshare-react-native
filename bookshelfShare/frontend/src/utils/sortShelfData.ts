export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
