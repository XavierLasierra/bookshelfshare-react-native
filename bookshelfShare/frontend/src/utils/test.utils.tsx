import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
