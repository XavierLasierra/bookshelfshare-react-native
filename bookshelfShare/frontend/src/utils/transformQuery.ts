interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
