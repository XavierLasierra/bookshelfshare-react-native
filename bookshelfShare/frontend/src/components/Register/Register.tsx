import React, { useState } from 'react';
import {
  View, Text, TextInput, TouchableOpacity, SafeAreaView
} from 'react-native';
import {
  SharedElement
} from 'react-navigation-shared-element';

import { useDispatch } from 'react-redux';
import { validateEmail, validatePassword } from '../../utils/validation.utils';
import { registerUser } from '../../redux/actions/loggedUser.creator';

import BookIcon from '../../assets/bookIcon.svg';
import ArrowIcon from '../../assets/arrowIcon.svg';
import globalStyles from '../../styles/global.styles';
import styles from './register.styles';

export default function Register({ navigation: { pop } }:any) {
  const dispatch = useDispatch();
  const [username, setUsername] = useState('');
  const [isValidUsername, setValidUsername] = useState(true);
  const [email, setEmail] = useState('');
  const [isValidEmail, setValidEmail] = useState(true);
  const [password, setPassword] = useState('');
  const [isValidPassword, setValidPassword] = useState(true);
  const [passwordRepeat, setPasswordRepeat] = useState('');
  const [isValidPasswordRepeat, setValidPasswordRepeat] = useState(true);

  function handleLoginNavigation() {
    pop();
  }

  function handleUsernameChange(text: string) {
    setUsername(text);
  }

  function handleUsernameFocus() {
    if (!isValidUsername) {
      setValidUsername(true);
    }
  }

  function handleEmailChange(text: string) {
    setEmail(text);
  }

  function handleEmailFocus() {
    if (!isValidEmail) {
      setValidEmail(true);
    }
  }

  function handlePasswordChange(text: string) {
    setPassword(text);
  }

  function handlePasswordFocus() {
    if (!isValidPassword) {
      setValidPassword(true);
    }
  }

  function handlePasswordRepeatChange(text: string) {
    setPasswordRepeat(text);
  }

  function handlePasswordRepeatFocus() {
    if (!isValidPassword) {
      setValidPasswordRepeat(true);
    }
  }

  function handleRegister() {
    if (!username) {
      return setValidUsername(false);
    }
    if (!validateEmail(email)) {
      return setValidEmail(false);
    }
    if (!validatePassword(password)) {
      return setValidPassword(false);
    }
    if (passwordRepeat !== password) {
      return setValidPasswordRepeat(false);
    }
    dispatch(registerUser({ username, email, password }));
    return pop();
  }

  return (
    <SafeAreaView style={styles.mainContainer}>
      <TouchableOpacity
        style={styles.backButton}
        onPress={handleLoginNavigation}
      >
        <ArrowIcon width={30} height={30} style={styles.backButtonIcon} />
      </TouchableOpacity>
      <View style={styles.topContainer}>
        <SharedElement id="mainIcon">
          <BookIcon
            width={50}
            height={50}
          />
        </SharedElement>
      </View>
      <View style={styles.registerContainer}>
        <Text style={globalStyles.title}>Sign up</Text>
        <View style={globalStyles.inputContainer}>
          <Text style={globalStyles.inputLabel}>username</Text>
          <TextInput
            style={globalStyles.input}
            value={username}
            onChangeText={handleUsernameChange}
            onFocus={handleUsernameFocus}
          />
          {!isValidUsername && <Text style={globalStyles.invalid}>Type your username</Text>}
        </View>
        <View style={globalStyles.inputContainer}>
          <Text style={globalStyles.inputLabel}>email</Text>
          <TextInput
            style={globalStyles.input}
            value={email}
            onChangeText={handleEmailChange}
            onFocus={handleEmailFocus}

          />
          {!isValidEmail && <Text style={globalStyles.invalid}>Invalid email</Text>}
        </View>
        <View style={globalStyles.inputContainer}>
          <Text style={globalStyles.inputLabel}>password</Text>
          <TextInput
            secureTextEntry
            style={globalStyles.input}
            value={password}
            onChangeText={handlePasswordChange}
            onFocus={handlePasswordFocus}
          />
          {!isValidPassword && (
          <Text style={[globalStyles.invalid, styles.invalidPasswordText]}>
            Minimum: 8 characters, 1 uppercase,
            1 lowercase and 1 number
          </Text>
          )}
        </View>
        <View style={globalStyles.inputContainer}>
          <Text style={globalStyles.inputLabel}>repeat password</Text>
          <TextInput
            secureTextEntry
            style={globalStyles.input}
            value={passwordRepeat}
            onChangeText={handlePasswordRepeatChange}
            onFocus={handlePasswordRepeatFocus}
          />
          {!isValidPasswordRepeat
          && <Text style={globalStyles.invalid}>Passwords don&quot;t match</Text>}
        </View>
        <TouchableOpacity
          style={[globalStyles.button, styles.registerButton]}
          onPress={handleRegister}
        >
          <Text style={globalStyles.buttonText}>Sign up</Text>
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  );
}
