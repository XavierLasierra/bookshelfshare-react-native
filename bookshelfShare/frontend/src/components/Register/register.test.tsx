import React from 'react';
import Register from './Register';
import { render, fireEvent } from '../../utils/test.utils';
import { registerUser } from '../../redux/actions/loggedUser.creator';

jest.mock('../../redux/actions/loggedUser.creator', () => ({
  registerUser: jest.fn()
}));

const navigation = {
  pop: jest.fn()
};

describe('Given a Register component', () => {
  describe('When it is rendered', () => {
    let screen: any;
    beforeEach(() => {
      screen = render(<Register navigation={navigation} />);
    });
    test('Then screen should match the snapshot', () => {
      expect(screen).toMatchSnapshot();
    });

    describe('And you click on loginPageButton', () => {
      test('Then navigation.pop should have been called', () => {
        const loginPageButton = screen.getByTestId('loginPageButton');
        fireEvent.press(loginPageButton);
        expect(navigation.pop).toHaveBeenCalled();
      });
    });

    describe('And usernameInput is empty', () => {
      describe('And you click registerButton', () => {
        beforeEach(() => {
          const registerButton = screen.getByTestId('registerButton');
          fireEvent.press(registerButton);
        });

        test('Then should render a Text with testId invalidUsername', () => {
          expect(screen.queryByTestId('invalidUsername')).not.toBe(null);
        });

        describe('And you focus on emailInput', () => {
          test('Then should not render a Text with testId invalidUsername', () => {
            const usernameInput = screen.getByTestId('usernameInput');
            fireEvent(usernameInput, 'focus');
            expect(screen.queryByTestId('invalidUsername')).toBe(null);
          });
        });
      });
    });

    describe('And username is valid but email is not', () => {
      describe('And you click registerButton', () => {
        beforeEach(() => {
          const usernameInput = screen.getByTestId('usernameInput');
          fireEvent(usernameInput, 'focus');
          fireEvent.changeText(usernameInput, 'Juan');
          const registerButton = screen.getByTestId('registerButton');
          fireEvent.press(registerButton);
        });

        test('Then should render a Text with testId invalidEmail', () => {
          expect(screen.queryByTestId('invalidEmail')).not.toBe(null);
        });

        describe('And you focus on emailInput', () => {
          test('Then should not render a Text with testId invalidEmail', () => {
            const emailInput = screen.getByTestId('emailInput');
            fireEvent(emailInput, 'focus');
            expect(screen.queryByTestId('invalidEmail')).toBe(null);
          });
        });
      });
    });

    describe('And username, email are valid but email is not', () => {
      describe('And you click registerButton', () => {
        beforeEach(() => {
          const usernameInput = screen.getByTestId('usernameInput');
          const emailInput = screen.getByTestId('emailInput');
          fireEvent.changeText(usernameInput, 'Juan');
          fireEvent(emailInput, 'focus');
          fireEvent.changeText(emailInput, 'juan@email.com');
          const registerButton = screen.getByTestId('registerButton');
          fireEvent.press(registerButton);
        });

        test('Then should render a Text with testId invalidPassword', () => {
          expect(screen.queryByTestId('invalidPassword')).not.toBe(null);
        });

        describe('And you focus on passwordInput', () => {
          test('Then should not render a Text with testId invalidPassword', () => {
            const passwordInput = screen.getByTestId('passwordInput');
            fireEvent(passwordInput, 'focus');
            expect(screen.queryByTestId('invalidPassword')).toBe(null);
          });
        });
      });
    });

    describe('And username, email, password are valid but passwordRepeat is not', () => {
      describe('And you click registerButton', () => {
        beforeEach(() => {
          const usernameInput = screen.getByTestId('usernameInput');
          const emailInput = screen.getByTestId('emailInput');
          const passwordInput = screen.getByTestId('passwordInput');
          fireEvent.changeText(usernameInput, 'Juan');
          fireEvent.changeText(emailInput, 'juan@email.com');
          fireEvent(passwordInput, 'focus');
          fireEvent.changeText(passwordInput, 'Aa111111');
          const registerButton = screen.getByTestId('registerButton');
          fireEvent.press(registerButton);
        });

        test('Then should render a Text with testId invalidPasswordRepeat', () => {
          expect(screen.queryByTestId('invalidPasswordRepeat')).not.toBe(null);
        });

        describe('And you focus on passwordRepeatInput', () => {
          test('Then should not render a Text with testId invalidPasswordRepeat', () => {
            const passwordRepeatInput = screen.getByTestId('passwordRepeatInput');
            fireEvent(passwordRepeatInput, 'focus');
            expect(screen.queryByTestId('invalidPasswordRepeat')).toBe(null);
          });
        });
      });
    });

    describe('And username, email, password, repeatPassword are valid', () => {
      describe('And you click registerButton', () => {
        const username = 'Juan';
        const email = 'juan@email.com';
        const password = 'Aa111111';
        beforeEach(() => {
          (registerUser as jest.Mock).mockReturnValue({ type: '' });

          const usernameInput = screen.getByTestId('usernameInput');
          const emailInput = screen.getByTestId('emailInput');
          const passwordInput = screen.getByTestId('passwordInput');
          const passwordRepeatInput = screen.getByTestId('passwordRepeatInput');
          fireEvent.changeText(usernameInput, username);
          fireEvent.changeText(emailInput, email);
          fireEvent.changeText(passwordInput, password);
          fireEvent(passwordRepeatInput, 'focus');
          fireEvent.changeText(passwordRepeatInput, password);
          const registerButton = screen.getByTestId('registerButton');
          fireEvent.press(registerButton);
        });

        test('Then should call registerUser with the username, email and password typed', () => {
          expect(registerUser).toHaveBeenCalledWith({ username, email, password });
        });

        test('Then should call navigation.pop', () => {
          expect(navigation.pop).toHaveBeenCalled();
        });
      });
    });
  });
});
