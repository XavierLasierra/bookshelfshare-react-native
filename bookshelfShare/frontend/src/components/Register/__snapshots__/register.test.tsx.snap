// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Register component When it is rendered Then screen should match the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#D9D8C4",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 20,
        "opacity": 1,
        "position": "absolute",
        "top": 35,
        "zIndex": 1,
      }
    }
    testID="loginPageButton"
  >
    <IconMock
      height={30}
      style={
        Object {
          "transform": Array [
            Object {
              "rotateZ": "180deg",
            },
          ],
        }
      }
      width={30}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 100,
        "justifyContent": "center",
        "paddingHorizontal": 30,
      }
    }
  >
    <View
      collapsable={false}
    >
      <IconMock
        height={50}
        width={50}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5F5F5",
        "borderTopRightRadius": 40,
        "display": "flex",
        "flexDirection": "column",
        "height": 600,
        "justifyContent": "space-between",
        "padding": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3E3C37",
          "fontSize": 35,
          "fontWeight": "500",
        }
      }
    >
      Sign up
    </Text>
    <View
      style={
        Object {
          "height": 70,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "transform": Array [
              Object {
                "translateX": 5,
              },
              Object {
                "translateY": 20,
              },
            ],
            "zIndex": 1,
          }
        }
      >
        username
      </Text>
      <TextInput
        onChangeText={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#ECECEC",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": "100%",
            "paddingHorizontal": 30,
            "paddingTop": 30,
          }
        }
        testID="usernameInput"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "transform": Array [
              Object {
                "translateX": 5,
              },
              Object {
                "translateY": 20,
              },
            ],
            "zIndex": 1,
          }
        }
      >
        email
      </Text>
      <TextInput
        onChangeText={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#ECECEC",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": "100%",
            "paddingHorizontal": 30,
            "paddingTop": 30,
          }
        }
        testID="emailInput"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "transform": Array [
              Object {
                "translateX": 5,
              },
              Object {
                "translateY": 20,
              },
            ],
            "zIndex": 1,
          }
        }
      >
        password
      </Text>
      <TextInput
        onChangeText={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#ECECEC",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": "100%",
            "paddingHorizontal": 30,
            "paddingTop": 30,
          }
        }
        testID="passwordInput"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "bold",
            "transform": Array [
              Object {
                "translateX": 5,
              },
              Object {
                "translateY": 20,
              },
            ],
            "zIndex": 1,
          }
        }
      >
        repeat password
      </Text>
      <TextInput
        onChangeText={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#ECECEC",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": "100%",
            "paddingHorizontal": 30,
            "paddingTop": 30,
          }
        }
        testID="passwordRepeatInput"
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#000000",
          "borderRadius": 10,
          "display": "flex",
          "height": 70,
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="registerButton"
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
            "fontWeight": "500",
          }
        }
      >
        Sign up
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
